### ISSUE TITLE
Add ARIA attributes for better accessibility

### ISSUE THREAD

**Issue by basham** [gpbl/react-day-picker#33]
While `react-day-picker` works with the keyboard, its semantics could be improved by adding the appropriate ARIA attributes, so screen readers will better interpret how the component works.

Probably the best React exemplar I've found which achieves this nicely is [`react-widgets` Calendar](http://jquense.github.io/react-widgets/docs/#/calendar). However, I see the customizability of the `react-day-picker` project being much more ideal. I really like the clean, small ES6 implementation with no dependencies outside of React.

The following are some suggestions based on my research. Some suggestions are a result of adding back certain semantics which were lost due to [switching from `<table>` to `<div>`](https://github.com/gpbl/react-day-picker/issues/25). Suggestions are organized by the element to which it affects, identified by its class name.
## `.DayPicker-Month`
- Add [`role="grid"` attribute](http://www.w3.org/TR/wai-aria/roles#grid); grid cells need an ancestor `grid`.

> Authors MUST ensure that elements with role gridcell are owned by elements with role row, which in turn are owned by an element with role rowgroup, grid or treegrid.
- Add [`aria-readonly="true"` attribute](http://www.w3.org/TR/wai-aria/states_and_properties#aria-readonly) to indicate that the child grid cells will never be edited; by default, grid cells are assumed to be editable.

> A grid is considered editable unless otherwise specified. To make a grid read-only, set the aria-readonly attribute of the grid to true. The value of the grid element's aria-readonly attribute is implicitly propagated to all of its owned gridcell elements, and will be exposed through the accessibility API.
- Add [`aria-labelledby` attribute](http://www.w3.org/TR/wai-aria/states_and_properties#aria-labelledby) to explicitly indicate the `.DayPicker-Caption` element is labeling the month grid.
- Add [`aria-activedescendant` attribute](http://www.w3.org/TR/wai-aria/states_and_properties#aria-activedescendant) to explicitly indicate which `.DayPicker-Day` is in focus.
## `.DayPicker-Caption`
- Add unique id to work with `.DayPicker-Month` enhancements.
## `.DayPicker-Weekdays`
- Add [`role="rowgroup"` attribute](http://www.w3.org/TR/wai-aria/roles#rowgroup), which is equivalent to using `<thead>` element.
## `.DayPicker-Weekday`
- Add [`role="columnheader"` attribute](http://www.w3.org/TR/wai-aria/roles#columnheader), which is equivalent to using the `<th>` element.
## `.DayPicker-Body`
- Add [`role="rowgroup"` attribute](http://www.w3.org/TR/wai-aria/roles#rowgroup), which is equivalent to using `<tbody>` element.
## `.DayPicker-Day`
- Add unique id to work with `.DayPicker-Month` enhancements.
- Add [`aria-disabled="true|false"` attribute](http://www.w3.org/TR/wai-aria/states_and_properties#aria-disabled) to indicate if the cell is selectable; this suggests that not every day within a given range of dates has to be selectable.
- Add [`aria-selected="true|false"` attribute](http://www.w3.org/TR/wai-aria/states_and_properties#aria-selected) to indicate if the cell is selected.

> gridcell elements with the aria-selected attribute set can be selected for user interaction
- Add [`title="Today"` attribute](http://www.w3.org/TR/html401/struct/global.html#adef-title) to a day which is today; if not wanting the tooltip to appear, then use [`aria-label="Today"` attribute](http://www.w3.org/TR/wai-aria/states_and_properties#aria-label). Alternatively, this suggestion could be implemented via a custom day component, rather than in this project.
## Additional considerations

Since `aria-labelledby` and `aria-activedescendant` reference an id, it may be good to dynamically generate unique ids for elements they reference, so you avoid the problem of having different instances of `react-day-picker` accidentally cross-reference each other.

**Comment by gpbl** [comment]
@basham thanks for taking the time to write this, it is really helpful 👍. Many suggestions are trivial to implement (see unit test below). 

For supporting `aria-selected` and `aria-disabled` I need to treat `selected` and `disabled` as "special" modifiers: I'll open a separate issue for this. 

`title` or `aria-label` attributes raise a problem with localization. "Today" is a default modifier but developers may want to implement other labels, such as "Bookable" or "Free" or whatever. I need to think about it, maybe adding `labels` prop...

```js
let captionElements = getElementsByClassName(dayPickerNode, "DayPicker-Caption");
let monthElements = getElementsByClassName(dayPickerNode, "DayPicker-Month");
let weekdaysElements = getElementsByClassName(dayPickerNode, "DayPicker-Weekdays");
let weekdayElements = getElementsByClassName(dayPickerNode, "DayPicker-Weekday");
let bodyElements = getElementsByClassName(dayPickerNode, "DayPicker-Body");
let weekElements = getElementsByClassName(dayPickerNode, "DayPicker-Week");
let dayElements = getElementsByClassName(dayPickerNode, "DayPicker-Day");

expect(dayPickerNode.getAttribute("role")).to.equal("widget");
expect(monthElements.every(el => el.getAttribute("role") === "grid")).to.be.true;
expect(monthElements.every(el => el.getAttribute("aria-readonly"))).to.equal("true");

monthElements.forEach((el, i) => {
  let captionId = captionElements[i].getAttribute("id");
  expect(el.getAttribute("aria-labelledby")).to.equal(captionId);
});

expect(weekdaysElements.every(el => el.getAttribute("role"))).to.equal("rowgroup");
expect(weekdayElements.every(el => el.getAttribute("role") === "columnheader")).to.be.true;
expect(bodyElements.every(el => el.getAttribute("role") === "rowgroup")).to.be.true;
expect(weekElements.every(el => el.getAttribute("role") === "row")).to.be.true;
expect(dayElements.every(el => el.getAttribute("role") === "gridcell")).to.be.true;
```

**Comment by basham** [comment]
Figured most of these suggestions would be really easy to implement. Thanks for the quick work on this.

I also assumed that `aria-disabled`, `aria-selected`, `title` and `aria-label` would need some additional work and experimentation. Glad to see those issues are getting special attention.

**Comment by basham** [comment]
I retract my suggestion of having `aria-activedescendant` on `.DayPicker-Month`. Turns out that approach would be most appropriate if you would leave focus only on `.DayPicker[role="widget"]` and it would never move elsewhere in the widget.

Since you're already managing focus on widget descendants, it's probably best to just leave the code as-is. Additionally, focus is a more established feature than `aria-activedescendant`, so what you've already done is a more backwards compatible approach.

## References

- [3.2.2. Using activedescendant to Manage Focus for Widget Children](http://www.w3.org/WAI/PF/aria-practices/#focus_activedescendant)
- [3.2.3. Managing Visual Focus with tabindex Alone](http://www.w3.org/WAI/PF/aria-practices/#visualfocus)

**Comment by ezufelt** [comment]
Would be happy to test a PR with a screen-reader.

**Comment by davidgilbertson** [comment]
I'm sure you know already, but you can easily test the compliance here: https://validator.w3.org/nu/#textarea

At the moment it's the only thing stopping our site from passing WCAG AA, so looking forward to this one!

(BTW, great component)

**Comment by gpbl** [comment]
Hi @davidgilbertson yes thank you for the link! 
I won't forget this issue... I'm just a bit overwhelmed by the quantity of things i've to read and pay attention to :-)

**Comment by davidgilbertson** [comment]
If I am able to get this change set up as a task in my project, would you consider a pull request?

**Comment by gpbl** [comment]
I would really appreciate any help, @davidgilbertson!

**Comment by davidgilbertson** [comment]
OK it looks like I can get this as a task in my next sprint. To be clear I plan to fix these three issues thrown by the w3 validator:
* Bad value `widget` for attribute `role` on element `div`.
* Element `attr` not allowed as child of element `div` in this context.
* An element with `role=row` must be contained in, or owned by, an element with `role=treegrid` or `role=grid` or `role=rowgroup`.

Should come up a few weeks from now.

**Comment by davidgilbertson** [comment]
I'm interested, what is `<attr>` [in the source](https://github.com/gpbl/react-day-picker/blob/1f4911b4bcadd0689ed8ceef175cb46a36efdeeb/src/DayPicker.js#L359)? I've never seen that before and can't find any info on it.

**Comment by gpbl** [comment]
OMG @davidgilbertson I think that `attr` was meant `abbr` 😭 fixed with #106! Great you've found it!

**Comment by davidgilbertson** [comment]
Happens to the best of us. :)

**Comment by ragesoss** [comment]
Hi! This is just to say that my project has really appreciate this react-day-picker, and we're looking forward to improvements in accessibility. One of my colleagues uses a screen reader, and cannot use the calendar in our app. There are, in particular, two issues for screen readers:

* Day cells are not clickable
    * We solved that problem on a derivative "weekday picker" by changing `div`s to `button`s,: https://github.com/ntdb/react-weekday-picker/commit/64d1be8b54a24ff94b757f4b9769f3cdc81d442e
* There is no indication of selected/not-selected/disabled state.
    * This is what adding the `aria-selected` attributes would do.

Echoing @ezufelt above, I'm happy to help test any patches that might fix those issues.

**Comment by gpbl** [comment]
Thanks @ragesoss! There's [a PR](https://github.com/gpbl/react-day-picker/pull/132) from @limscoder that may add some improvements for ARIA, but I'm not sure if it fixes all the issues. Going to check it out soon.

**Comment by davidgilbertson** [comment]
FWIW...

After thinking about this for a while, if I couldn't see (or was a keyboard navigator), I would much rather just type a date than navigate a calendar, so we are aria-hiding the calendar and just having an input field to type date.

**Comment by ragesoss** [comment]
@davidgilbertson That makes sense for picking a single date — I know that's what my colleague Helaine prefers when it's a matter of one date — but will it work for a calendar that is being used to select multiple dates? In my app, we use it with three states and lots of selected dates at once.
![day picker](https://cloud.githubusercontent.com/assets/848483/13441489/4202ab9c-dfac-11e5-905a-f3c7f2ed4317.png)

**Comment by gpbl** [comment]
@ragesoss is right saying that for multiple days input fields would not be enough.
I think we can't escape it, this component must support users adopting assistive technologies. 

There are already three pull requests (two of them are rather old) waiting to be analyzed and merged. It seems, however, they require a good amount of work :sweat_smile: I'm committed to work again on this soon!
